rasa/core/policies/policy.py

```diff
@@ -399,7 +400,7 @@ def __init__(
         policy_priority: int = 1,
         events: Optional[List[Event]] = None,
         optional_events: Optional[List[Event]] = None,
-        is_end_to_end_prediction: bool = False,
+        is_end_to_end_prediction: Optional[bool] = False,
```

Comment:
I like the limited scoped of the fix but I don't like the additional meaning in the boolean value which isn't obvious immediately (the meaning of `is_end_to_end_prediction=None` only becomes obvious by looking onto more code / comments). I'd prefer one of these options
- adding another field `is_loop_prediction` to `PolicyPrediction`
- adding a field `prediction_type` to `PolicyPrediction` with an `Enum` type `PredictionType`. This is more scalable if we add further prioritization of special cases later (e.g. for fallbacks (just an example)) but would require a deprecation of  `is_end_to_end_prediction` which makes it slightly more work.

Reply:
effectively we have `3` types of prediction, regardless of loop of fallback, etc: `without using user input`, `with user intent`, `with user text`. The reason I like reusing the same bool, as both `without using user input` and `with user intent` require adding `DefinePrevUserUtteredFeaturization(False)`. Since it is all about the user I prefer current way, since it precisely represents that, where `None` mean that we don't know

Reply:
@joejuzl @alwx Any strong opinion? I'd still favor explicitness as it makes the logic easier to understand without too much further background.

Reply:
Do we only ever convert it to a boolean? Or do we ever branch on it being `None`

Reply:
in `ensemble.py`, we branch if it is `None`

Reply:
@Ghostvv Considering [Joe's comment](https://github.com/RasaHQ/rasa/pull/7749#discussion_r563863122) I'd gone with one of my two solutions then . Do you expect further prediction types or or do you expect it to stay these 3 for a while?

Reply:
as you know, it's hard to predict 🙈 But I think we can go with three for now

Reply:
I'll update

Reply:
I added the unnecessary `bool`